### PR TITLE
Move data logic out of the inserter

### LINKED ISSUE #4221: "Reusable Block" shouldn't show in the inserter
image

### LINKED ISSUE #4303: Use the underlying block icon for reusable blocks in the inserter
Currently, reusable blocks appear in the inserter with a mosaic icon:

screen shot 2018-01-05 at 09 47 31

It would be nicer if they appeared with the icon of the block that was converted:

screen shot 2018-01-05 at 09 46 44

### PR DESCRIPTION
**Pull request by noisysocks**
### 💃 What 

Fixes #4221. Fixes #4303.

Currently, `InserterMenu`, `VisualEditorInserter`, and `blocksAutocompleter` all directly call `getBlockTypes()` and `getReusableBlocks()` to determine what blocks should be shown.

This means that we duplicate the same logic in a few different places:

* Checking, if the block has `useOnce` set, whether it's been already used or not
* Checking whether or not the block has `isPrivate` set
* Checking whether or not the block is in the editor config's list of enabled block types
* Mixing in the list of available reusable blocks with the list of regular blocks

This approach is pretty error-prone: it's very easy to change the behaviour of one inserter and forget about another.

To start addressing this, I've changed `InserterMenu` to source its data from two new selectors:

* `getInserterItems` - used to populate the inserter's search results
* `getRecentInserterItems` - used to populate the 'Recent' tab

Both selectors filter out private/disabled blocks, correctly set `isDisabled`, and are unit tested.

### ✅ How to test

1. Insert some static blocks and reusable blocks
2. Recently inserted blocks should move to the top of the Recent tab in the main inserter

### CODE REVIEW

File: editor/components/inserter/group.js [hunk 1]
```diff
@@ -10,71 +10,92 @@ import { Component } from '@wordpress/element';
...
+	 * @inheritdoc
+	 */
 	componentWillReceiveProps( nextProps ) {
-		if ( ! isEqual( this.props.blockTypes, nextProps.blockTypes ) ) {
-			this.activeBlocks = deriveActiveBlocks( nextProps.blockTypes );
+		if ( ! isEqual( this.props.items, nextProps.items ) ) {
+			this.activeItems = deriveActiveItems( nextProps.items );
 			// Try and preserve any still valid selected state.
-			const current = find( this.activeBlocks, { name: this.state.current } );
+			const current = find( this.activeItems, ( item ) => isEqual( item, this.state.current ) );
```
Comment (youknowriad) [inline editor/components/inserter/group.js]: Can we avoid `isEqual` and use natural keys instead? (performance reasons)

File: editor/components/inserter/group.js [hunk 2]
```diff
@@ -10,71 +10,92 @@ import { Component } from '@wordpress/element';
...
+		const { onSelectItem } = this.props;
 
 		return (
 			<button
 				role="menuitem"
-				key={ block.name === 'core/block' && block.initialAttributes ?
-					block.name + block.initialAttributes.ref :
-					block.name
-				}
+				key={ index }
```
Comment (youknowriad) [inline editor/components/inserter/group.js]: Not sure it's a good idea to use `index` here. I'd prefer a natural key as well. The block index can change when searching for instance.

File: editor/store/selectors.js [hunk 3]
```diff
@@ -1111,15 +1111,129 @@ export function getNotices( state ) {
...
+	const staticItems = getBlockTypes().map( blockType =>
+		buildInserterItemFromBlockType( state, enabledBlockTypes, blockType )
+	);
+
+	const dynamicItems = getReusableBlocks( state ).map( reusableBlock =>
+		buildInserterItemFromReusableBlock( enabledBlockTypes, reusableBlock )
+	);
+
+	const items = [ ...staticItems, ...dynamicItems ];
+	return compact( items );
```
Comment (youknowriad) [inline editor/store/selectors.js]: Can we add a unique `id` to each item. It could equal the `name` for regular blocks and the `ref` for reusable blocks. Having a unique id across all items would simplify logic in the inserter.

**Review by youknowriad — COMMENTED** [review]
Nice work, One concern left for me (the unique id) and this should be good to go.

File: editor/components/inserter/group.js (hunk 1 above)
Comment (noisysocks) [reply editor/components/inserter/group.js]: 👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.

File: editor/components/inserter/group.js (hunk 2 above)
Comment (noisysocks) [reply editor/components/inserter/group.js]: 👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.

File: editor/store/selectors.js (hunk 3 above)
Comment (noisysocks) [reply editor/store/selectors.js]: 👍 fixed in ef3bbc97f96bc9c5516d7501c4039fc7a70dc47b.

File: editor/store/selectors.js [hunk 4]
```diff
@@ -1111,15 +1111,131 @@ export function getNotices( state ) {
...
 /**
- * Resolves the list of recently used block names into a list of block type settings.
- *
- * @param {Object} state Global application state
- *
- * @returns {Array} List of recently used blocks.
+ * An item that appears in the inserter. Inserting this item will create a new
+ * block. Inserter items encapsulate both regular blocks and reusable blocks.
+ * 
+ * @typedef {Object} Editor.InserterItem
```
Comment (youknowriad) [inline editor/store/selectors.js]: We're missing the id here ;)

**Comment by youknowriad** [comment]
Do you plan to use these in the other insereters in a separate PR?

This looking solid for me. (Text are broken though for some reason).

File: editor/store/selectors.js (hunk 4 above)
Comment (noisysocks) [reply editor/store/selectors.js]: Whoops, fixed :)

**Comment by noisysocks** [comment]
> Do you plan to use these in the other inserters in a separate PR?

Yep, I plan to use these methods in the `blocksAutocompleter`. There's no need to update `VisualEditorInserter` since it's going away.

File: editor/store/selectors.js [hunk 5]
```diff
@@ -1111,15 +1111,132 @@ export function getNotices( state ) {
...
+ * @param {string[]|boolean}      enabledBlockTypes Enabled block types, or true/false to enable/disable all types.
+ * @param {Object}                blockType         Block type, likely from getBlockType().
+ * @returns {Editor.InserterItem}                   Item that appears in inserter.
+ */
+function buildInserterItemFromBlockType( state, enabledBlockTypes, blockType ) {
+	if ( ! enabledBlockTypes || ! blockType ) {
+		return null;
+	}
+
+	const blockTypeIsDisabled = Array.isArray( enabledBlockTypes ) && ! enabledBlockTypes.includes( blockType.name );
```
Comment (aduth) [inline editor/store/selectors.js]: ES6 prototype methods are not polyfilled (e.g. `Array.from` is fine, but `Array#includes` is not).

This will error in IE11.

Comment (noisysocks) [reply editor/store/selectors.js]: 👌 good to know. Looks like https://github.com/WordPress/gutenberg/pull/5528 addresses this.

Comment (noisysocks) [reply editor/store/selectors.js]: Could we polyfill prototype methods?

Comment (aduth) [reply editor/store/selectors.js]: Context at #746

tl;dr: Was waiting for Babel 7, which I didn't expect to take so long to release 😄
